dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/ConnectorArguments.java

```diff
@@ -816,21 +825,67 @@ public String getDefinition(@Nonnull ConnectorProperty property) {
     }
     String result = definitionMap.get(property.getName());
     LOG.info(
-        "Retrieving {} from key {} - got: {}",
-        property.getName(),
+        "Retrieving {} from key '{}' - got: '{}'",
         property.getDescription(),
+        property.getName(),
         MoreObjects.firstNonNull(result, "[not found]"));
     return result;
   }
 
   public void searchDefinitions() {
-    definitionMap = new HashMap<>();
-    for (String definition : getOptions().valuesOf(definitionOption)) {
-      if (definition.contains("=")) {
-        int idx = definition.indexOf("=");
-        definitionMap.put(definition.substring(0, idx), definition.substring(idx + 1));
+    String connectorName = getConnectorName();
+    ImmutableSetMultimap<String, String> propertyNameToConnectors =
+        createPropertyNameToConnectorsMapping();
+    definitionMap =
+        getOptions().valuesOf(definitionOption).stream()
+            .map(
+                definition -> {
+                  int idx = definition.indexOf('=');
+                  String propertyName;
+                  Optional<String> propertyValue;
+                  if (idx >= 0) {
+                    propertyName = definition.substring(0, idx);
+                    propertyValue = Optional.of(definition.substring(idx + 1));
+                  } else {
+                    propertyName = definition;
+                    propertyValue = Optional.empty();
+                  }
+                  Preconditions.checkArgument(
+                      propertyNameToConnectors.containsKey(propertyName),
+                      "Unrecognized command-line option '-D%s'.",
+                      propertyName);
+                  ImmutableSet<String> compatibleConnectors =
+                      propertyNameToConnectors.get(propertyName);
+                  Preconditions.checkArgument(
+                      compatibleConnectors.contains(connectorName),
+                      "Command-line option '-D%s' cannot be used with connector '%s'. "
+                          + "Compatible connector(s): '%s'.",
+                      propertyName,
+                      connectorName,
+                      compatibleConnectors);
+                  return propertyValue
+                      .map(value -> immutableEntry(propertyName, value))
+                      .orElseThrow(
+                          () ->
+                              new IllegalArgumentException(
+                                  String.format(
+                                      "Command-line option '-D%s' requires a value. Use: -D%s=value",
+                                      propertyName, propertyName)));
```

Comment:
Is this a deferred exception which is always thrown if the value is empty? In which case, could we have done something above like:
```
if (idx < 0)
    throw new IAE(...)
```

instead of storing the Optional in the ImmutableMap and then using orElseThrow as a functional form here?